Lib/test/test_exceptions.py

```diff
@@ -1285,6 +1285,18 @@ def g():
                 next(i)
                 next(i)
 
+    def test_assert_shadowing(self):
+        # Shadowing AssertionError would cause the assert statement to
+        # misbehave.
+        globals()['AssertionError'] = TypeError
+        global AssertionError
+        try:
+            assert False, 'hello'
```

Comment:
This test will fail when run in optimized mode (`-O`).

Reply:
It does not fail, but it silently passes without testing anything. `test_assert_with_tuple_arg()` in test_raise.py tests `assert` in the same way. I added a `skipUnless` decorator.

Reply:
You can add also a test for
```python
exec(compile('assert False, "hello"', 'test', 'exec', optimize=0), {'AssertionError': TypeError})
```
which can work even in optimized mode.

Reply:
@serhiy-storchaka Should that line go in its own test method?

Reply:
If you prefer.